### PR TITLE
Footstep sounds

### PR DESCRIPTION
**Pull request by tigercat2000**
This adds a system for footstep sounds
 - Only xenomorphs and humans currently use it, could expand to robutts
   and simple_animals easily, but haven't.
 - Only enabled for three turf types currently:
  - Plating (ala Maintnance)
  - Carpets
  - Wooden floors

Demo video (someone yell at me if the values are fucked, I tried my best to make them noticable yet quiet) https://puu.sh/l5Xj5/3e93ae3b04.mp4

### CODE REVIEW

**Comment by Fox-McCloud** [comment]
:-1: 

I've seen this system in action on servers in the past. It makes stealth impossible in a lot of scenarios once it's completely fleshed out. The person in maintenance who's spying on someone? Stealth ruined because of clank clank plating.

It's also really annoying to be inundated with incessant clanking/shuffling/creaking when you have a lot of people in a small space.

Even Bay, who's typically not typically as concerned with the impacts of gameplay mechanics, shot this down.

**Comment by tigercat2000** [comment]
@Fox-McCloud 
I can adjust the values much more than they already are, if it's such a problem. As is, unless the sound system is bugged, running with shoes will generate a 15 volume sound for an entire screen, and running without shoes should generate a 7 volume sound for half of a screen.

A suggested item for traitors was stealthy shoes, which prevent footstep noises- Or, perhaps, it could be applied to the no slip shoes, I'm not sure.

It only affects a few turfs as-is, exactly because hearing footsteps all the time while wandering around on the normal station tiles would be annoying as fuck, especially when many people are gathered together- Given the relative rarity of plating, carpets, and wooden floors, I think that is more or less fine.

But the major point you have is that people can actually tell where people are going- I don't consider that a problem. It's certainly a change to the mechanics, but I don't think it's necessarily a _bad_ change.

**Comment by Fox-McCloud** [comment]
I wouldn't consider having plating make noise as being "a few turf types". 99% of maintenance is plating. It's like adding it only to "floors" and saying "it's only one turf type".

The problem is, stealth has never been about avoid how you walk or having to have specific clothing. Likewise, not everyone has noslip shoes or access to them (changelings, revs, cult, shadowlings, etc). 

Furthermore, even if they did, they're still having to invest those extra resources JUST for the ability to walk with stealth...not to mention it makes it BLATANTLY obvious who has traitor shoes on, as well "oh, he's syndicate; I saw him walking over carpet/plating and he was silent" (yes, this WILL happen, it already does when someone with shoes walks over banana peels). 

Basically take the clomp clomp of security and add it to everyone; that's the effects on behavior it's going to have--anddd I don't think that's a good thing.

**Comment by tigercat2000** [comment]
@Fox-McCloud 
Would this be better with a much-reduced range? Say, walking only makes noise 1 tile in every direction around you, running 2-3? Or perhaps quieter than 15 and 7?

I'm not especially after fucking over stealth- I mostly just like the idea of having some kind of noise to walking. I think being able to notice people in the darkness from that alone is interesting, but to each their own.

**Comment by KasparoVy** [comment]
I feel like dropping the range and the volume a bit would make this viable. It wouldn't mean people'd have to try that much harder to stealth. Not everyone's a ninja

**Comment by Spacemanspark** [comment]
:+1: if the range is decreased.

**Comment by Regen1** [comment]
:+1: 
This adds a lot to atmosphere and opens up for more traitor-y things
Instead of disregarding this as a feature because "oh muh gameplais" why not think more of "How can we expand on this".
For an example, sneaky loafers which makes no sound when walking on plating?
Syndicate sneaky loafers which makes no sound disguised as normal boots (patent pending)

**Comment by Fox-McCloud** [comment]
@Regen1 

How can you say this opens up more traitor shennanigans when this objectively makes traitoring harder? Forcing traitors to spend more of their TC just to avoid detection is NOT "opening up" things for traitors, it's straight up nerfing them for no apparent reason.

Likewise, as I pointed out, SS13 isn't just about traitors, it's about a plethora of game-modes, and only traitors would have access to the tools you mentioned; everyone else just straight up gets screwed over. And again, it makes its SUPERRRR easy to meta people because if their shoes are slient, you know for absolute certainty they're a traitor.

**Comment by Plotron** [comment]
If the range is decreased to about 3 tiles, what is the point (beyond hearing in the dark)? What about the annoyance factor?
I think Fox is right here.

Sure it may be infuriating if you lose a track of somebody in maintenance, but I think we've got a perfect balance here (damned walkways!).

**Comment by tigercat2000** [comment]
I really don't give a fuck about the balancing aspect of this. I'm happy to play along with it to try to get it through, but I really don't give a single fuck about losing people in maintnance or "oh no the stealth is impossible!", I like the idea of this for the possibilities it opens up, and the ambiance.

**Comment by Aurorablade** [comment]
i agree with a range/volume reduction on things.

**Comment by tigercat2000** [comment]
Already working on it, as well as some other additions, just taking a while due to (fucking complex math)&trade;

**Comment by tigercat2000** [comment]
From the commit
Changes:
 - Humans now use the following ranges:
   - range(6) for running + shoes
   - range(5) for running OR shoes
   - range(4) for not running AND no shoes
 - Humans use the following volumes:
   - 13 for running + shoes
   - 9 for running OR shoes
   - 5 for not running AND no shoes
 - Humans have special modifiers on steps:
   - Shoes being 'silent' = Silent
     - Noslip shoes are silent by default.
     - You can make any shoe silent by tablecrafting shoe wraps with 10
     ducttape and a wirecutter at a table.
   - Species being 'silent' = Silent
     - Currently, shadowlings use this.
   - Buckled, lying, or being thrown = Silent
   - No feet = Silent
   - No gravity = Footsteps every 3 turfs instead of every 2.

 - Xenomorphs now use the following ranges:
   - range(4) at all times.
 - Xenomorphs use the following volumes:
   - 5 at all times.
 - Xenomorphs have special modifiers on steps:
   - Walking = Silent
   - Buckled, lying, being thrown = Silent
   - No gravity = Footsteps every 3 turfs instead of every 2.

**Comment by tigercat2000** [comment]
Please note, `range()` refers to BYOND's insane get_dist system. Basically, 6 = 7, counting from 1.
0/| is center, person hearing footsteps.

```
      6
     656
    65456
   6543456
  654323456
 65432123456
654321|123456
 65432123456
  654323456
   6543456
    65456
	 656
	  6
```

**Comment by Earthdivine** [comment]
I don't see where "my stealth and balance" comes into play when you already can see the entire screen in the first place, and if your hidden in darkness, then that adds a further play in your making it out to be. Run away if their chasing you, duck in a corner and stop. The noise stopped, does this mean they got away or are hiding? There's very little stealth element at all, specifically due to how view range works.

However, these ranges are really loud, in terms of comparing it to life. I certainly don't hear the footsteps of someone a small ways down the hallway unless they have heavy footfalls. It makes me feel like everyone's wearing mag boots when imagining how it would be.

**Comment by tigercat2000** [comment]
@Earthdivine I dunno, I can try to tune it down even further (oh god not more ComplicatedMath&trade;), but, I don't think it's really a big problem, they are rather quiet compared to other sound making things.

**Comment by Crazylemon64** [comment]
I find that playing with headphones on, I'm able to hear sound most other players seem to ignore. So, volume might not be the best way to balance this.

**Comment by tigercat2000** [comment]
@Crazylemon64 I wear headphones as well, yet I have to fuck with my sound settings and set it to precisely far-too-fucking-loud levels to hear anything with SS13. So I dunno. My point however was that it does make less noise than any other thing in the game as far as I am aware.

**Comment by Earthdivine** [comment]
Running looks good range wise, walking you shouldn't hear people much at all, even with wearing shoes. 

Though, to be honest I also get the feelin this will go the way Text2Speech did. You'll be hearing this sound about 30+ times, I can't think of a way this wouldn't be super annoying, as cool and atmospheric as it is.

**Comment by tigercat2000** [comment]
@Earthdivine I dunno, I don't think it'll be too bad with only happening in maintenance, the few areas that have wooden floors, and on the few carpets scattered around the station.

**Comment by KasparoVy** [comment]
Would species known for their above-par hearing benefit from this system in any way? Or does it just treat everyone as human? Like, how there is a night-sight flag, there could be a stepheard-dist flag or something. *Shrug*.

I could only imagine a species with ears as large as the tajaran or vulpkanin, or those simply as sensitive as the vox, would be able to benefit atleast somewhat-- but idk, muh balance??

**Comment by tigercat2000** [comment]
@KasparoVy More of, it currently uses playsound(), which treats all mobs equally in terms of sound, so I couldn't really make a species advantage anyways.

**Comment by FreeStylaLT** [comment]
:+1: the atmosphereseses yesyes

I approve of pretty much any iteration of this as long as tap tap tap is heard

**Comment by Robertha89** [comment]
if ya can hear it through walls, then nay.

**Comment by tigercat2000** [comment]
@Robertha89 Why do you consider that a big problem?

**Comment by Robertha89** [comment]
cause the walls are thick, and usually there to hide people, knowing where people are from sound, while it sounds like a good thing, is not good due to walking, if the sound was based on the area you are in, instead of like this, it would be fine, because then you could hear someone on another end, which is also the reason every shoe does not have walking noises to begin with. I like the idea, but not it's current iteration, as being able to tell where people are based on where they walk, i mean this is how you know security is near, or clown is near.

Do also take in consideration, NOSLIP flag is also applied to magboots, so magboots are super silent.

**Comment by tigercat2000** [comment]
@Robertha89 They aren't, I was referring specifically to the syndicate no-slip shoes, they now have a variable on their path that makes them silent.

As-is, the sound will transfer through walls to some degree- The exact degree, I don't know, I have absolutely no idea if the sound system takes walls into account to dampen noise. 
To actually hear someone in maintenance realistically, you'd both have to be pressed up against an adjoining wall. Not to mention, all they have to do is walk past rooms or take off their shoes to get an instant range reduction to it.

As for areas, that wouldn't make any more sense, (especially as I have no fucking idea how the hell that would work), because you would have people running around literally next to you, but due to the area, not hear them.

**Comment by Robertha89** [comment]
to do the proper echo sound, you would need a seperate system to handle that, which would be outside the scope of this pr, oh, and you should add a toggle footsteps button to preferences, and maybe convert the security boots/sandals and clown shoes too the same same system, for consistency.

**Comment by tigercat2000** [comment]
@Robertha89 
Adding it to preferences is not only impossible with how it currently works, but also an incredibly bad idea. The only thing worse than the game mechanics changing is changing to something inconsistant. Either people can turn off footsteps for themselves, in which case it becomes willingly nerf vs. keep it the same, or they can turn off hearing footsteps, in which case people have no idea who can hear them.

Security boots and clown shoes work just fine as they are, this system doesn't currently have support for shoe-based sounds, only turfs.

**Comment by Robertha89** [comment]
In which case, you should allow people to decide if they can hear it, or put it under Ambience.

**Comment by tigercat2000** [comment]
Ugh, fuck this, whatever, fuck this entire system, fuck arguing with people, and most importantly
# FUCK EVER TRYING TO CHANGE SOMETHING

**Comment by tigercat2000** [comment]
And now, I am going to take a nap.

**Comment by Spacemanspark** [comment]
I really did like it, @tigercat2000

**Comment by Aurorablade** [comment]
@tigercat2000 ....if you reopen and keep working on it i will port over something of your choice....or offer you a kitten...

**Comment by tigercat2000** [comment]
@Spacemanspark @Aurorablade  It's an uphill fucking battle I don't want to fight. If it's not someone who saw a similar system and hated it, it's someone who wants to balance it before it can ever see any practical testing. It's not worth fucking arguing over.

**Comment by Earthdivine** [comment]
And a potential cool thing is lost due to balance.

**Comment by Aurorablade** [comment]
@tigercat2000 *holds up kitten..makes sad eyes...*

**Comment by Aurorablade** [comment]
@Earthdivine it wasn't even balance it was silly nitpicking...the option thing would only be valid just to make it a challange mode for the player to not know who can hear them. I am almost tempted to grab this so it dosen't die..but i am not good at super complicated mathnumbering...i mean i could also bribe @tigercat200 with porting over the meat spike thing...but i think they just need a nap and to be buried in kittens..i have a box of them, help me bury...

**Comment by Spacemanspark** [comment]
"Balance" strikes again.

**Comment by FreeStylaLT** [comment]
pls don't close

**Comment by Aurorablade** [comment]
*passes out boxes of fluffy kittens to throw at @tigercat2000 *

**Comment by Spacemanspark** [comment]
!(Shitgoeshere)[https://www.lovethisgif.com/uploaded_images/733-hey-open-it.gif]

**Comment by TheDZD** [comment]
Honestly, it seems like people really overblew the balance aspects of this feature. Creepy footsteps in maintenance would've been pretty great.

**Comment by orelbon** [comment]
why dont you just make footsteps go away when walking, like if you where sneaking.

**Comment by CrAzYPiLoT-SS13** [comment]
Damn,  tiger said exactly the thing that happens to almost every feature
PR. I liked this. RIP.

**Comment by Mijhead** [comment]
and again, meaningless balance has taken precedence over actual fun.

**Comment by Spacemanspark** [comment]
9-10 people want this.
3-4 people are complaining about it.
Can we at least try it out?

**Comment by tigercat2000** [comment]
Right, I had a nice nap. I guess I can reopen this, but I have to note a few things first:
 - I **really** don't want to try fucking with the range any more. It took nearly 2 hours to get the damn `range(6), range(5), range(4)` stuff working, and to make it start off even smaller would probably take another 2 hours.
 - I'm not adding a preference option to turn it off, because that is just *begging* for problems, "HE METAED ME!" "no you have footsteps off he has them on", or "he turned footsteps off so you didn't hear him"

File: code/modules/mob/living/carbon/alien/alien.dm [hunk 1]
```diff
@@ -295,3 +295,30 @@ Des: Removes all infected images from the alien.
...
+				var/volume = 5
+
+				if(m_intent == "walk")
+					return 0 //silent when walking
+
+				if(buckled || lying || throwing)
+					return 0 //people flying, lying down or sitting do not step
+
+				if(!has_gravity(src))
+					if(step_count % 3) //this basically says, every three moves make a noise
```
Comment (RemieRichards) [inline code/modules/mob/living/carbon/alien/alien.dm]: Not what this says at all.
1 mod 3 is 1
2 mod 3 is 2
3 mod 3 is 0

```if(step_count % 3 == 0)``` is "every 3 steps"

**Comment by RemieRichards** [comment]
You've written your ```%``` backwards.
They work as you intend, but they're written oddly.

**Comment by tigercat2000** [comment]
@RemieRichards Have I? I think most of our `%` operators are formatted `var % num`

**Comment by RemieRichards** [comment]
@tigercat2000 it's mainly that, in your conditions, they pass the majority of the time (which leads to a return, so an "end") and fail the least of the time (which leads to the sound, a "start"

So your code is largely "end","end","start",
Logically, you'd sync up the fail/success of the conditional with the outcome, eg: a passing if lead to a successful sound. plus the ```%``` operation you're using CHECKS for those failures, rather than say, ```if(steps % 3 == 0) playsound()```

I'm rambling, but it just makes your modulo conditionals a little less readable.

**Comment by Aurorablade** [comment]
also to address the issue @Earthdivine or..someone brought up about 30 sounds at once. Is there anyway to do a check of people in range of you and just queue out/limit the number of sounds like..30 people in one spot..only play like 5...yeah it cuts people out but it avoids the issue..in my head this works. As i said before..this often never leads to good things.

**Comment by tigercat2000** [comment]
@Aurorablade I could do something like
```
if >5 mobs in view
 if prob 50
   return 0
```
or something

**Comment by Plotron** [comment]
People mostly seem to want this feature because it's new, it's fancy and it adds something to the game.
Fox says he's seen it in practice and it's not worth it.

Sure, if you so insist, give it a try and see if it changes anything.
(PS, I like the feature that lets you craft silent shoes)

**Comment by Regen1** [comment]
Yeah to be completely honest I want to throw the "BUT MUH BALANCE" bullshit straight out the fucking window on this one.
If this really gets some people metagame alarm screaming then it'd be as easy as to make craftable silent loafers.

**Comment by orelbon** [comment]
@tigercat2000  @Fox-McCloud sorry for the noob question but how do i use a mapmerger?

**Comment by tigercat2000** [comment]
@orelbon go to tools in the downloaded repo, then mapmerge, hit 'prepare_maps.bat', apply your changes, run 'clean_maps.bat', type in the file names you changed (relative to map_files, ex, if you modify cyberiad.dmm, enter 'cyberiad/cyberiad.dmm')

**Comment by orelbon** [comment]
@tigercat2000 thanks!

**Comment by orelbon** [comment]
@tigercat2000 something is wrong its not letting me i get this error 
error

**Comment by tigercat2000** [comment]
@orelbon You didn't point to the cyberiad file, you need to point to the cyberiad file like `cyberiad/cyberiad.dmm`

**Comment by orelbon** [comment]
is this ok?
ok

**Comment by tigercat2000** [comment]
@orelbon provided you did actually prepare_maps before editting the map, it looks fine.

**Comment by orelbon** [comment]
@tigercat2000 Hoe do you place an object on a wall that you only want to be accessed from one room, as in so it doesn't get accessed from the other side.

**Comment by tigercat2000** [comment]
@orelbon Place it on the turf next to a wall in the room you want it to be in, right click on it, hit edit, then change the pixel_x and pixel_y values until it is on the wall.

**Comment by stobarico** [comment]
I have am idea how to fix the antag problem, when walking, the sounds do not play (stalking when xeno) when running or hunting, it plays the sounds, that would make sneaking better (kinda) and we would have the sounds

**Comment by orelbon** [comment]
@tigercat2000 could you explain how to update the fork i got from para because i have to delete the repository and fork it every time i want to make a new PR.

**Comment by tigercat2000** [comment]
@orelbon I don't know what you are using for git.
The github client has a pull or update button, I think.
Git's command line would be `git fetch upstream` followed by `git reset --hard upstream/master`

**Comment by orelbon** [comment]
@tigercat2000 i use the standard github desktop app

**Comment by tigercat2000** [comment]
@orelbon probably this then https://help.github.com/desktop/guides/contributing/syncing-your-branch/

**Comment by orelbon** [comment]
@tigercat2000 im pretty lost i cant understand what they want you to do.

**Comment by tigercat2000** [comment]
@orelbon I have no idea, I don't use the github desktop client.

**Comment by orelbon** [comment]
@tigercat2000 tiger when placng blast doors do you have to connect them to a button or do you just have to get the correct button?

**Comment by tigercat2000** [comment]
@orelbon you have to set the id_tag on the door and the id on the button

**Comment by orelbon** [comment]
@tigercat2000  how do you set who has access to use a button?

**Comment by tigercat2000** [comment]
@orelbon no idea, try setting req_access_txt

**Comment by Regen1** [comment]
Merge when?

**Comment by Fox-McCloud** [comment]
Will be reviewing and testing the intricacies of this as soon as I can get to it; sorry for the delay.

**Comment by Fox-McCloud** [comment]
@Fox-McCloud

Test this, locally.

**Comment by orelbon** [comment]
enough people seem to want this, why not merge it and monitor feedback on it to see if it really works or not?

**Comment by tigercat2000** [comment]
29 days whoo
I think this might break the rigsuit record for "most delayed PR"!

**Comment by orelbon** [comment]
5 years later.....

**Comment by tigercat2000** [comment]
36 days now!

Rigsuits was 48 days.

**Comment by Spacemanspark** [comment]
Can this just get merged already.

**Comment by orelbon** [comment]
^^^Can we at least test it out to see if people like it or not.

**Comment by Regen1** [comment]
Merge, this is in favor of merging.

**Comment by ZomgPonies** [comment]
Worth testing :+1: . This has been in favour of being merged for a LONG time now.

**Comment by orelbon** [comment]
so whse gonna do the honorary merge?

**Comment by Fox-McCloud** [comment]
@tigercat2000 

Having actually tested this, some things to note:

- The woodstepping is a bit loud compared to plating; it sounds like wooden clogs walking on the floor or something.
- Even if you're not moving, the sounds will still trigger as long as you're holding an arrow key (this shouldn't happen, IMO, and jackboots/clown shoes don't follow this logic).
- the carpet sound is imperceptible, though that may be a good thing.
- the sound falloff seems ok
- shoe silencing works fine.

Personally, I'd equalize the sounds a bit more--tone down the plating a bit and the wood a bit and bump up the carpet just a hair (probably not much; it's carpet afterall).

**Comment by tigercat2000** [comment]
@Fox-McCloud 
About the moving thing, nothing I can do.
It's called in Move(), which is part of byond's automatic shit, I can't fix byond being trash and calling it when the movement fails. No clue why shoes don't do the exact same.

**Comment by Fox-McCloud** [comment]
Will re-test this tomorrow, and likely merge it then.

**Comment by Spacemanspark** [comment]
YES.

**Comment by M3hillus** [comment]
Dont want to step over any toes (ha) but is there a potential, eventally of course, where walking is quieter than running?

**Comment by tigercat2000** [comment]
walking already is quiter than running @M3hillus

**Comment by TheDZD** [comment]
Tested this locally a bit, sound seems alright unless I have a lot of mobs walking around near each other and generating a lot of sounds (hooray for mutilating code in order to make mobs walk around in circles really fast).

**Comment by orelbon** [comment]
YES

**Comment by tigercat2000** [comment]
dickbutt

**Comment by FalseIncarnate** [comment]
:+1: for that gif.
